KAFKA-19020 [1/n]: Handle strict max fetch records in share fetch

JimmyWang6 opened the pull request:
JIRA:
[https://issues.apache.org/jira/projects/KAFKA/issues/KAFKA-19020](https://issues.apache.org/jira/projects/KAFKA/issues/KAFKA-19020)
KIP :
[here](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1206:+Strict+max+fetch+records+in+share+fetch)
The `maxFetchRecords` parameter in shared fetch operations currently
acts as a soft limit, meaning the actual number of records returned to
the client may exceed this threshold under certain conditions. This PR
introduces an `AcquireMode` in the `ShareFetchRequest` to offer clients
a more flexible method for acquiring records. By setting this value to
strict, consumers can ensure that the actual number of records received
will never exceed the limit specified by the max.poll.records
configuration.

Reviewers: Apoorv Mittal <apoorvmittal10@gmail.com>, Lan Ding
 <isDing_L@163.com>, Abhinav Dixit <adixit@confluent.io>, Andrew
 Schofield <aschofield@confluent.io>

JimmyWang6 commented:
Hi @apoorvmittal10, I've opened a draft PR to initiate a discussion about KIP-19020.
In the current approach, when the maxFetchRecords limit is exceeded, we'll use `firstAcquiredOffset + maxFetchRecords - 1` to determine the offset instead of the batch's last offset. I've conducted local tests, and it appears to be work as expected.
Could you have a look when available? thanks.

apoorvmittal10 commented:
@JimmyWang6 I think we should support both i.e. current and strict functionality. That means if the acquire method is called with acquire mode as `strict` then we should only acquire till `maxFetchRecords` else can go upto batch boundaries for optimization. This will give SharePartition flexibility to support both.

From clients, we need to have another param in ShareFetchRequest which governs if the max records are strict. The client changes can be done later as it will require a small KIP, cc: @ShivsundarR.

JimmyWang6 commented:
@apoorvmittal10 
Sorry for the late response. I think you mean to add an extra ` acquireMode` field for shareFetch request and allowing clients to explicitly choose between strict mode or loose mode (current implementation).
I will draft a KIP to formalize this change later. Thanks for the feedback.

apoorvmittal10 commented:
Yes, it will need a minor KIP for RPC and client side change. But we can implement the functionality in server sider i.e. in share partition by changing the `acquire` method to support both functionalities earlier. Becuase this will unblock the other issue we have https://issues.apache.org/jira/browse/KAFKA-17541. In case you think this is something someone else or I should take up, then feel free to re-assign to myself.

JimmyWang6 commented:
Sorry again for the delay. Initially, I didn't see the best way forward, and it left me a bit stuck.
Since I know the assignee of KAFKA-17541 , I'm all for having a discussion together to move these two issues forward.
I haved started drafting the KIP [KIP-1206](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1206%3A+Strict+max+fetch+records+in+share+fetch#KIP1206:Strictmaxfetchrecordsinsharefetch-Compatibility,Deprecation,andMigrationPlan) but maybe I'll need some guidance. Thanks for your help!

JimmyWang6 commented:
@apoorvmittal10 I've completed the KIP for this issue. I would greatly appreciate it if you could take a moment to review [this](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1206:+Strict+max+fetch+records+in+share+fetch) in advance and let me know if there are any areas that need improvement or if anything is missing. After that, I'll start a discussion thread.

JimmyWang6 committed:
KAFKA-19020: Handle strict max fetch records in share fetch

# Conflicts:
#	clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java

JimmyWang6 committed:
KAFKA-19020: Handle strict max fetch records in share fetch

# Conflicts:
#	clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java

JimmyWang6 committed:
KAFKA-19020: Handle strict max fetch records in share fetch

JimmyWang6 committed:
KAFKA-19020: Handle strict max fetch records in share fetch

apoorvmittal10 commented on clients/src/main/resources/common/message/ShareFetchRequest.json:
```diff
@@ -32,6 +33,8 @@
       "about": "The current share session epoch: 0 to open a share session; -1 to close it; otherwise increments for consecutive requests." },
     { "name": "MaxWaitMs", "type": "int32", "versions": "0+",
       "about": "The maximum time in milliseconds to wait for the response." },
+    { "name": "AcquireMode", "type": "string", "versions": "2+",
+      "about": "The acquire mode to control max fetch records behavior" },
```
Shall we please remove client side changes from the PR and let's have only server side changes as that not requires a KIP.

JimmyWang6 replied:
Copy that. I have reverted the related client side changes.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/ShareAcquireMode.java:
```diff
@@ -0,0 +1,37 @@
+
+public enum ShareAcquireMode {
+    STRICT("STRICT"), 
+    LOOSE("LOOSE");
```
May be the better word would be `BATCH_OPTIMIZED`, @AndrewJSchofield wdyt?

AndrewJSchofield replied:
Yes, I agree.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -668,6 +668,7 @@ public long nextFetchOffset() {
     @SuppressWarnings({"cyclomaticcomplexity", "methodlength"}) // Consider refactoring to avoid suppression
     public ShareAcquiredRecords acquire(
         String memberId,
+        String acquireMode,
```
As we are not certain about client side changes for now because that might have higher egress for clients given we have to anyways send the complete batch but with partial acquired records hence we should not have the `acquireMode` in the `acquire` API for now. Rather we can decide dynamically regarding how the behaviour should be for any fetch.

JimmyWang6 replied:
Done.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -746,7 +747,7 @@ public ShareAcquiredRecords acquire(
                 // Do not send the lastOffsetToAcquire as when the subMap is empty, it means that
                 // there isn't any overlap itself.
-                ShareAcquiredRecords shareAcquiredRecords = acquireNewBatchRecords(memberId, fetchPartitionData.records.batches(),
+                ShareAcquiredRecords shareAcquiredRecords = acquireNewBatchRecords(memberId, acquireMode, fetchPartitionData.records.batches(),
```
We should default to `BATCH_OPTIMIZED/LOOSE` acquire mode in this method and can solve dynamic toggle in the ticket: https://issues.apache.org/jira/browse/KAFKA-17541

JimmyWang6 replied:
Done. I have changed the acquire mode to `BATCH_OPTIMIZED/RECORD_LIMIT` according to @AndrewJSchofield comments.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1553,12 +1555,15 @@ private ShareAcquiredRecords acquireNewBatchRecords(
             long lastAcquiredOffset = lastOffset;
+            long maxOffset = firstAcquiredOffset + maxFetchRecords - 1;
             if (maxFetchRecords < lastAcquiredOffset - firstAcquiredOffset + 1) {
-                // The max records to acquire is less than the complete available batches hence
-                // limit the acquired records. The last offset shall be the batches last offset
-                // which falls under the max records limit. As the max fetch records is the soft
-                // limit, the last offset can be higher than the max records.
```
Rather removing the comment, enhance it with the changes.

JimmyWang6 replied:
Done.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2311,6 +2316,24 @@ long findLastOffsetAcknowledged() {
         return lastOffsetAcknowledged;
     }
 
+    /**
+     * Find the maximum offset from the first batch within the request offset for {@link ShareAcquireMode.STRICT} mode.
+     *
+     * @param batches The batches to search for the request offset.
+     * @param offset The request offset to find.
+     * @return The minimum value between the first batch's last offset and the request offset.
+     *         If no batches are available, returns the original request offset.
+     */
+    private long maxOffsetFromFirstBatch(
+        Iterable<? extends RecordBatch> batches,
+        long offset
+    ) {
+        for (RecordBatch batch : batches) {
+            return Math.min(batch.lastOffset(), offset);
+        }
+        return offset;
+    }
```
Can you please help clarify regarding what we want to achieve here? Seems there is a `for` loop but the comparison will happen only for first batch, then why do we need a `for` loop? Also say the batches are [0-2],[3-5] i.e. 2 batches with 3 records each. And in strict mode the maxFetchRecords are 4 then what batches will be acquired? Do we have a test case for this?

JimmyWang6 replied:
I have changed the code structure and polished the comments, and 2 test cases are added. As for 2 batches with 3 records each when maxFetchRecords are 4, only 3 records in first batch will returned. I will add more unit tests if this mechanism looks good to you :)

apoorvmittal10 replied:
This still seems a problem.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1553,12 +1555,15 @@ private ShareAcquiredRecords acquireNewBatchRecords(
-                // limit, the last offset can be higher than the max records.
-                lastAcquiredOffset = lastOffsetFromBatchWithRequestOffset(batches, firstAcquiredOffset + maxFetchRecords - 1);
+                ShareAcquireMode mode = acquireMode(acquireMode);
+                lastAcquiredOffset = switch (mode) {
+                    case STRICT ->
+                        maxOffsetFromFirstBatch(batches, maxOffset);
```
This seems to be incorrect as we ll now create partial batches in SharePartition, not aligned on batch boundaries. We should still create batches on batch boundaries but should acquire only subset.

JimmyWang6 replied:
Hi @apoorvmittal10 , thanks for your coments!
Do you mean that the `lastAcquiredOffset` should still be aligned on boundaries?
Although partial batches are created here, the actual records returned from the server by invoking `kafka.server.share.ShareFetchUtils#maybeSliceFetchRecords` are still aligned.

The lastAcquiredOffset here is used for two purposes:

1. Updating the endOffset for SharePartition.
2. Setting the lastOffset value for AcquiredRecords.

IMHO, I don't think the firstOffset and lastOffset should be strictly aligned to batch boundaries. I have considered the following scenario:

1. A producer sends 1000 records in **a single batch**.
2. 10 share-consumers are started with `share.max.poll.records` or `max.poll.records` set to 1.

The expected behavior should be that each consumer receives 1 record per poll() call, which is the current behavior. WDYT?

apoorvmittal10 replied:
> Do you mean that the lastAcquiredOffset should still be aligned on boundaries?

Yes, that's the default assumption in the code.

> I have considered the following scenario:
A producer sends 1000 records in a single batch.
10 share-consumers are started with share.max.poll.records or max.poll.records set to 1.

So a batch of 1000 records with 0-999 offsets (say that's the offsets for the batch were), then a single batch of 0-999 should be created. But the offsets should be acquired as per the strict limit i.e. if only 1 record is requested then only 0 offset should be acquired. Aligning to batch boundaries is simple and that's the assumption in the code as well while doing other operations.

apoorvmittal10 reviewed (changes requested):
Thanks for the PR, I did the initial review. Please find some comments.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2311,6 +2321,24 @@ long findLastOffsetAcknowledged() {
     }
 
+    /**
+     * In STRICT mode, only acquire records from the first batch.
```
Why it should be the behaviour that only records from first batch should be acquired?

apoorvmittal10 reviewed (commented):
Thanks for the changes, left couple of comments to align the code design first.

JimmyWang6 commented:
Hi @apoorvmittal10,
I have updated the PR which can work in my local machine, please help to check if this is the right way to implement this when you are free, thanks.

DL1231 commented:
@JimmyWang6 Thanks for the patch, please fix the CI error.

DL1231 reviewed (commented):
@JimmyWang6 Thanks for the patch, please fix the CI error.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -675,6 +675,9 @@ public ShareAcquiredRecords acquire(
         FetchIsolation isolationLevel
     ) {
         log.trace("Received acquire request for share partition: {}-{} memberId: {}", groupId, topicIdPartition, memberId);
+        // Since the client side changes are not yet in place, use BATCH_OPTIMIZED as the default AcquireMode.
+        ShareAcquireMode acquireMode = acquireMode(ShareAcquireMode.BATCH_OPTIMIZED.toString());
+        boolean strict = ShareAcquireMode.STRICT.equals(acquireMode);
```
Thinking again regardin how it will be used until we have capability from client to tell the acquire mode.

For broker, the acquire method should determine if strict fetch is needed. To determine same it could be based on delivery count i.e. the fetchOffset is available in cache and delivery count is x. So we do not need `ShareAcquireMode` class at all. Just a function call like below, for now:

```
boolean strict = isStrictAcquireMode()
```
```
boolean isStrictAcquireMode() {
    return false;
````

Later, we will implement  `isStrictAcquireMode()` to return value based on some inputs.

JimmyWang6 replied:
@apoorvmittal10 Thank you for your comments. I have made the suggested changes and added some more tests. Please take a look when you have time :)

apoorvmittal10 reviewed (commented):
Thanks for the changes, can you please include the tests as well for the change. I have let a comment for now, will review once again though.

JimmyWang6 committed:
add client side changes

JimmyWang6 committed:
add client side changes

JimmyWang6 committed:
add client side changes

JimmyWang6 committed:
add client side changes

JimmyWang6 committed:
fix build

JimmyWang6 committed:
fix build

JimmyWang6 committed:
server side changes

JimmyWang6 committed:
server side changes

JimmyWang6 committed:
add ShareAcquireMode field to ShareFetchRequest

JimmyWang6 committed:
add ShareAcquireMode field to ShareFetchRequest

JimmyWang6 committed:
Merge branch 'KAFKA-19020' of https://github.com/JimmyWang6/jimmy-KAFKA-14048 into KAFKA-19020

JimmyWang6 committed:
polish client side changes

JimmyWang6 commented:
Hi @apoorvmittal10 @AndrewJSchofield ，
Since the KIP has been approved in the vote, I believe I can proceed with the implementation. I’ve prepared an initial version of the framework and some tests, could you help to review them for any major issues? Thanks in advance.

AndrewJSchofield commented:
> Hi @apoorvmittal10 @AndrewJSchofield ， Since the KIP has been approved in the vote, I believe I can proceed with the implementation. I’ve prepared an initial version of the framework and some tests, could you help to review them for any major issues? Thanks in advance.

@JimmyWang6 Thanks for the PR. I will review it but I observe that this is quite a large PR which means it's going to take a while to review. It's going to take approvals from both @apoorvmittal10 and me before we can merge.

JimmyWang6 commented:
@AndrewJSchofield  Noted. Thanks for your reply.

JimmyWang6 committed:
merge trunk

JimmyWang6 committed:
merge trunk

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -817,9 +832,13 @@ public ShareAcquiredRecords acquire(
                     // all the records from the batch as the batch will anyway be part of the file-records
                     // response batch.
-                    int acquiredSubsetCount = acquireSubsetBatchRecords(memberId, firstBatch.baseOffset(), lastOffsetToAcquire, inFlightBatch, result);
+                    int acquiredSubsetCount = acquireSubsetBatchRecords(memberId, isRecordLimitMode, maxFetchRecords, firstBatch.baseOffset(), lastOffsetToAcquire, inFlightBatch, result);
```
We would want to change this comment now since we are sending `maxFetchRecords` to the method

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1709,6 +1780,10 @@ private int acquireSubsetBatchRecords(
                     .setLastOffset(offsetState.getKey())
                     .setDeliveryCount((short) offsetState.getValue().deliveryCount()));
                 acquiredCount++;
+                if (isRecordLimitMode && acquiredCount >= maxFetchRecords) {
```
should the condition `acquiredCount > maxFetchRecords` be allowed, I think it should only be `acquiredCount == maxFetchRecords`

JimmyWang6 replied:
@adixitconfluent Thanks for your comments! You are right, and I was just trying to be more defensive here.

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1596,9 +1628,45 @@ private ShareAcquiredRecords acquireNewBatchRecords(
+            CompletableFuture<Void> future = new CompletableFuture<>();
+            NavigableMap<Long, InFlightState> offsetStateMap = inFlightBatch.offsetState();
+            for (Map.Entry<Long, InFlightState> offsetState : offsetStateMap.tailMap(lastOffset, false).entrySet()) {
+                if (offsetState.getValue().state() == RecordState.ACQUIRED) {
```
1. I guess there should be some logging for the `else` corresponding to this `if`
2. Question: Do we need to adjust the `endOffset` in this function?

JimmyWang6 replied:
1. Done.
2. I don't think we need to adjust it. The `endOffset` should remain aligned with the batch's boundary and avoid moving backward since it's already set to `lastAcquiredOffset` in `acquireNewBatchRecords`. However, I'm happy to reconsider if you have a different perspective.

adixitconfluent replied:
I think I am happy with the current approach. Just want to highlight this comment for @apoorvmittal10 to take a look at as well.

apoorvmittal10 replied:
I am sorry but not aligned with the current approach. I have suggested an alternative.

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1596,9 +1628,45 @@ private ShareAcquiredRecords acquireNewBatchRecords(
         }
     }
 
+    private AcquiredRecords filterShareAcquiredRecordsInRecordLimitMode(int maxFetchRecords, List<AcquiredRecords> acquiredRecords) {
```
This function should be within write lock

JimmyWang6 replied:
Yes, I have checked that all invocations of this function are within a write lock.

adixitconfluent replied:
you should either write a comment on top of this function that this function should be called only when the caller holds the write lock 
OR 
put the function within write lock
```
lock.writeLock().lock();
        try {
              ---
              ---
        } finally {
              lock.writeLock().unlock();
        }
```

We don't know how we are going to utilize this function in the future. So although, the code looks good right now, in the future it may create a problem.

JimmyWang6 replied:
I just got what you mean, done~

adixitconfluent reviewed (changes requested):
Thanks @JimmyWang6 for the PR. I have reviewed the broker source code. Left some comments.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareSessionHandler.java:
```diff
@@ -78,6 +79,11 @@ public class ShareSessionHandler {
+    /**
+     * The share acquire mode for this session.
+     */
+    private String shareAcquireMode;
```
Let's use the enum here instead of String.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -2639,6 +2640,7 @@ private <K, V> void buildRequestManager(MetricConfig metricConfig,
         int maxBytes = Integer.MAX_VALUE;
         int fetchSize = 1000;
         int minBytes = 1;
+        ShareAcquireMode shareAcquireMode = ShareAcquireMode.of("batch_optimized");
```
Can't you just use `ShareAcquireMode.,BATCH_OPTIMIZED`?

JimmyWang6 replied:
Thank you for the comment, let me address it.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareSessionHandlerTest.java:
```diff
@@ -54,6 +55,7 @@
 @Timeout(120)
 public class ShareSessionHandlerTest {
     private static final LogContext LOG_CONTEXT = new LogContext("[ShareSessionHandler]=");
+    private final ShareAcquireMode shareAcquireMode = ShareAcquireMode.of("batch_optimized");
```
I would expect at least one test using `ShareAcquireMode.RECORD_LIMIT` to make sure it gets set correctly. You've just tested the default so far.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -106,6 +108,7 @@ public class ShareConsumeRequestManager implements RequestManager, MemberStateLi
                                final ShareConsumerMetadata metadata,
                                final SubscriptionState subscriptions,
                                final FetchConfig fetchConfig,
```
I think it's time to make a `ShareFetchConfig` which copies all of the `FetchConfig` but also adds in `share.acquire.mode`. We are bound to add more configs over time and I don't like adding new parameters every time. A quick search of the code suggests that `ShareFetchConfig` would fit in cleanly (unlike `ShareSubscriptionState` which I'd love to create, but `SubscriptionState` is pervasive).

JimmyWang6 replied:
Noted. Initially, I wanted to add `share.acquire.mode` to `FetchConfig`, but then I realized it is also used by classic consumer groups.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareSessionHandler.java:
```diff
@@ -111,7 +117,8 @@ public boolean isNewSession() {
 
-    public ShareFetchRequest.Builder newShareFetchBuilder(String groupId, FetchConfig fetchConfig) {
+
+    public ShareFetchRequest.Builder newShareFetchBuilder(String groupId, ShareAcquireMode shareAcquireMode, FetchConfig fetchConfig) {
```
This is an example of why I'd like `ShareFetchConfig` instead.

JimmyWang6 replied:
Done.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. I've reviewed the client code and left some comments.

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -817,9 +832,13 @@ public ShareAcquiredRecords acquire(
+                    int acquiredSubsetCount = acquireSubsetBatchRecords(memberId, isRecordLimitMode, maxFetchRecords, firstBatch.baseOffset(), lastOffsetToAcquire, inFlightBatch, result);
                     acquiredCount += acquiredSubsetCount;
-                    continue;
+                    if (isRecordLimitMode) {
```
I am not sure why do we need this piece of code, can you please add some comments why we need it?

JimmyWang6 replied:
The initial purpose here was that in `record_limit` mode, only a single batch would be returned in ShareAcquiredRecords. But after reconsidering, I think you are right and this code could be removed.

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -835,23 +854,35 @@ public ShareAcquiredRecords acquire(
+                        List.of(new AcquiredRecords()
+                            .setFirstOffset(inFlightBatch.firstOffset())
+                            .setLastOffset(inFlightBatch.lastOffset())
+                            .setDeliveryCount((short) inFlightBatch.batchDeliveryCount())));
```
we should also update the acquisition lock timeout for this batch, or would it not be required?

JimmyWang6 replied:
Thanks for the comment. It may not be necessary, since in the `filterShareAcquiredRecordsInRecordLimitMode `function, the batch-level acquisition lock timeout will be canceled by `maybeInitializeOffsetStateUpdate()`, and the offset-level acquisition lock timeout will be triggered instead.

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -835,23 +854,35 @@ public ShareAcquiredRecords acquire(
             // Some of the request offsets are not found in the fetched batches. Acquire the
             // missing records as well.
-            if (acquiredCount < maxRecordsToAcquire && subMap.lastEntry().getValue().lastOffset() < lastOffsetToAcquire) {
+            if (acquiredCount < maxRecordsToAcquire && subMap.lastEntry().getValue().lastOffset() < lastOffsetToAcquire && !isRecordLimitMode) {
```
I did not get this change. Why are we not honouring `isRecordLimitMode` here?

JimmyWang6 replied:
Same reason to [https://github.com/apache/kafka/pull/20246#discussion_r2449247007](https://github.com/apache/kafka/pull/20246#discussion_r2449247007). However, upon reflection, I think this approach might introduce other problems, so I removed this code.

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1596,9 +1628,45 @@ private ShareAcquiredRecords acquireNewBatchRecords(
+        AcquiredRecords records = acquiredRecords.get(0);
+        InFlightBatch inFlightBatch = cachedState.get(records.firstOffset());
+        long lastOffset = records.firstOffset() + maxFetchRecords - 1;
+        if (inFlightBatch != null) {
```
there should be some logging for the `else` corresponding to this `if`

JimmyWang6 replied:
Good catch, I'll address it.

JimmyWang6 committed:
resolve conflicts

JimmyWang6 committed:
address comments

JimmyWang6 committed:
address comments

JimmyWang6 committed:
address comments & ut

adixitconfluent commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -8801,6 +8838,294 @@ public void testRecordArchivedWithWriteStateRPCFailure() throws InterruptedExcep
+            .withDefaultAcquisitionLockTimeoutMs(ACQUISITION_LOCK_TIMEOUT_MS)
+            .withPersister(persister)
+            .build());
+        Mockito.doReturn(true).when(sharePartition).isRecordLimitMode(Mockito.any());
```
This is a little conflicting since here you are saying that the acquire mode is `RECORD_LIMIT` but below in `acquire` method you are passing `BATCH_OPTIMIZED`. I would say its better to remove this line and in `acquire` method pass `ShareAcquireMode.RECORD_LIMIT`

JimmyWang6 replied:
Thank you for pointing out this issue. The reason is that previously, the KIP hadn't been approved yet, so we used a mock approach to determine the `acquireMode`. Now, this part of the code is no longer necessary, and I will remove it.

adixitconfluent commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -8801,6 +8838,294 @@ public void testRecordArchivedWithWriteStateRPCFailure() throws InterruptedExcep
+
+        assertArrayEquals(expectedAcquiredRecord(0, 1, 1).toArray(), acquiredRecordsList.toArray());
+        assertThrows(IllegalStateException.class, () -> sharePartition.cachedState().get(0L).batchAcquisitionLockTimeoutTask());
+        // There should be 3 timer tasks for 3 offsets.
```
The comment is incorrect

JimmyWang6 replied:
Done.

adixitconfluent reviewed (changes requested):
I have reviewed the broker side tests. Left some comments. Overall, I feel we should cover many more unit  test cases of various cases when acquire can occur. example - when the batch to be acquired is in between, the offsets to acquire are not continuous etc.

JimmyWang6 commented:
@adixitconfluent 
Thanks for your review. I'll keep adding more unit test cases to cover various scenarios.

JimmyWang6 committed:
address comments and more tests

JimmyWang6 committed:
fix nit

JimmyWang6 committed:
more tests

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1614,9 +1649,50 @@ private ShareAcquiredRecords acquireNewBatchRecords(
+                        offsetState.getValue().state(), offsetState.getKey(), inFlightBatch, groupId, topicIdPartition);
+                }
+            }
+            rollbackOrProcessStateUpdates(future, null, persisterBatches);
```
Hmm, when we convert any record's state from `AVAILABLE` to `ACQUIRED`, we do not persist it. So why do we need to send this info to persister when we are converting the records back from `ACQUIRED` to `AVAILABLE`? I think we should only convert the record's state from `ACQUIRED` to `AVAILABLE` without any rollback?

adixitconfluent commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -8801,6 +8838,543 @@ public void testRecordArchivedWithWriteStateRPCFailure() throws InterruptedExcep
+            .withDefaultAcquisitionLockTimeoutMs(ACQUISITION_LOCK_TIMEOUT_MS)
+            .withPersister(persister)
+            .build());
+        Mockito.doReturn(true).when(sharePartition).isRecordLimitMode(Mockito.any());
```
same comment as https://github.com/apache/kafka/pull/20246#discussion_r2450531664, we should get rid of this statement and send `ShareAcquireMode.RECORD_LIMIT `. Please make the change to other tests as well.

JimmyWang6 replied:
Finish.

adixitconfluent commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -8801,6 +8838,543 @@ public void testRecordArchivedWithWriteStateRPCFailure() throws InterruptedExcep
+
+        sharePartition.maybeInitialize();
+
+        // Creating 3 batches starting from 11, such that there is a natural gap from 26 to 30
```
I think the gap is between offsets 36 to 40, so the comment is incorrect

JimmyWang6 replied:
Done.

adixitconfluent commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -8801,6 +8838,543 @@ public void testRecordArchivedWithWriteStateRPCFailure() throws InterruptedExcep
+
+        GapWindow persisterReadResultGapWindow = sharePartition.persisterReadResultGapWindow();
+        assertNotNull(persisterReadResultGapWindow);
+        // Gap still exists from 36 to 50
```
comment should be `Gap still exists from 36 to 40`

JimmyWang6 replied:
Done.

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -846,27 +861,45 @@ public ShareAcquiredRecords acquire(
             if (acquiredCount < maxRecordsToAcquire && subMap.lastEntry().getValue().lastOffset() < lastOffsetToAcquire) {
                 log.trace("There exists another batch which needs to be acquired as well");
-                ShareAcquiredRecords shareAcquiredRecords = acquireNewBatchRecords(memberId, fetchPartitionData.records.batches(),
+                int numRecordsRemaining = maxRecordsToAcquire - acquiredCount;
```
Do we have any test case covering this scenario?

JimmyWang6 replied:
Yes, I think the `testAcquireMultipleRecordsWithOverlapAndNewBatchInRecordLimitMode` and `estAcquireCachedStateInitialGapOverlapsWithActualPartitionGapInRecordLimitMode` tests could cover this case.

adixitconfluent reviewed (changes requested):
Thanks @JimmyWang6 for the changes. Overall looks much better. Left a few comments for `SharePartition` and `SharePartitionTest` file. I feel we can do better in terms of testing - 
1. Unit tests - Some scenarios need to be tested - 
           a. when we need to acquire batch as well as subset of records in a different batch.
           b. when no of records available to acquire are lesser than the max fetch records in various permutations of cached state.
           c. when we have offsets in cached state that are in order of AVAILABLE --- ACKNOWLEDGED --- AVAILABLE and we need to traverse over batches/records of multiple InFlightBatch to calculate the records to be acquired.
2. We can add some integration tests in `ShareConsumerTest.java`

JimmyWang6 committed:
address comments

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java:
```diff
@@ -0,0 +1,78 @@
+
+    public final byte id;
+
+    ShareAcquireMode(final String name, byte id) {
```
Why do we have final modifier only for name, have the modifier for both or none.

JimmyWang6 replied:
Done.

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public String toString() {
+        return super.toString().toLowerCase(Locale.ROOT);
```
In Kafka generally a format is used for toString:

```
return "ShareAcquireMode{"  +
            "name=" + name +
            ", id=" + id +
            "}";

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/ShareAcquireMode.java:
```diff
@@ -0,0 +1,36 @@
+ */
+package org.apache.kafka.server.share.fetch;
+
+public enum ShareAcquireMode {
```
Why do we need another server side class and can't use same client ShareAcquireMode class?

JimmyWang6 replied:
I previously thought that the core module couldn't import client dependencies. I misunderstood this part.

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/ShareFetch.java:
```diff
@@ -63,6 +63,10 @@ public class ShareFetch {
+    /**
+     * The share acquire mode for the fetch request.
+     */
+    private final byte shareAcquireMode;
```
nit: Rather than byte we should have ShareAcquireMode here itself i.e. convert to enum here itself so we need no conversion again.

JimmyWang6 replied:
Thanks for your comment, done.

JimmyWang6 replied:
Done.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1614,9 +1649,55 @@ private ShareAcquiredRecords acquireNewBatchRecords(
+                    }
+                }
+                rollbackOrProcessStateUpdates(future, null, persisterBatches);
+                records.setLastOffset(lastOffset);
```
This doesn't seems right, why do you want to make a call to persister while acquiring records?

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -748,8 +752,13 @@ public ShareAcquiredRecords acquire(
+                    AcquiredRecords records = filterShareAcquiredRecordsInRecordLimitMode(maxRecordsToAcquire, shareAcquiredRecords.acquiredRecords());
+                    // Return only single batch of filtered records.
+                    shareAcquiredRecords = new ShareAcquiredRecords(List.of(records), maxRecordsToAcquire);
+                }
```
`acquireNewBatchRecords` will update cache and create acquistion lock timeouts. Then later we call to `filterShareAcquiredRecordsInRecordLimitMode` and update the states. Can't we do something like below:

1. Move below code to a separate method frome `createBatches`:
```
result.forEach(acquiredRecords -> {
                // Schedule acquisition lock timeout for the batch.
                AcquisitionLockTimerTask timerTask = scheduleAcquisitionLockTimeout(memberId, acquiredRecords.firstOffset(), acquiredRecords.lastOffset());
                // Add the new batch to the in-flight records along with the acquisition lock timeout task for the batch.
                cachedState.put(acquiredRecords.firstOffset(), new InFlightBatch(
                    timer,
                    time,
                    memberId,
                    acquiredRecords.firstOffset(),
                    acquiredRecords.lastOffset(),
                    RecordState.ACQUIRED,
                    1,
                    timerTask,
                    timeoutHandler,
                    sharePartitionMetrics));
                // Update the in-flight batch message count metrics for the share partition.
                sharePartitionMetrics.recordInFlightBatchMessageCount(acquiredRecords.lastOffset() - acquiredRecords.firstOffset() + 1);
            });
```
2. The result from `acquireNewBatchRecords` will let us know what all batches to be made in cache, at most.
3. If the mode is `batch_optimized` then call new metod else if `record_limit` is the mode then only keep first batch in result.
4. If the size of the `acquireNewBatchRecords` first batch is greater than `max fetch records` then instatiate OffsetState accordingly.

This approach will keep the cache clean with only required data in cache. Else you might also need to worry about other cache capacity issues as well. Also creating and remving acquisition lock timeouts again is not good.

JimmyWang6 replied:
Thanks for your detailed suggestion! I have changed accordingly, please take a look again :)

apoorvmittal10 reviewed (changes requested):
Thanks for the PR. We have made really good effort in implementing the changes. There are some suggestion, specially for SharePartition where I think we can do better.

JimmyWang6 committed:
address comments

JimmyWang6 committed:
fix nit

JimmyWang6 committed:
address comments and more test in ShareConsumerTests

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1656,10 +1712,54 @@ private List<AcquiredRecords> createBatches(
+                    acquiredRecords = filterShareAcquiredRecordsInRecordLimitMode(maxFetchRecords, inFlightBatch);
+                    sharePartitionMetrics.recordInFlightBatchMessageCount(
+                        acquiredRecords.lastOffset() - acquiredRecords.firstOffset() + 1);
+                    return List.of(acquiredRecords);
```
Rather than first creating `AcquisitionLockTimerTask` for complete batch and then cancelling same and creating new per offset seems not required. When we know that it's first time batch is being created then why not to attach individual offsets locks there itself i.e. create an inflight batch but with offset based tracking by default rather filtering later.

JimmyWang6 replied:
Good idea, I've already made the changes.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -846,25 +851,35 @@ public ShareAcquiredRecords acquire(
-                inFlightBatch.updateAcquisitionLockTimeout(acquisitionLockTimeoutTask);
+                long numRecordsAcquired = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
+                int numRecordsRemaining = maxRecordsToAcquire - acquiredCount;
+                if (isRecordLimitMode && (numRecordsAcquired > numRecordsRemaining)) {
```
Seems like a full batch acquisition cannot happen here, so should we not handle this situation as like partial batch acknowledgement done above in this method i.e.`if (!fullMatch || inFlightBatch.offsetState() != null)` should have additional condition?

JimmyWang6 replied:
I'm sorry, I don't quite follow your point in this change, could you please provide more details or an example?

apoorvmittal10 replied:
So we are in the part of the code where whole batch should be acquired, but we eventually acquire partial bath. Which happens in phases, 1st the complete batch and then filtering.

But we can change the approach, this part of code should not be executed as complete batch is not being acquired rather partital hence the situation is not other than above where we check `(!fullMatch || inFlightBatch.offsetState() != null)`, hence we need to add another check i.e. `(!fullMatch || inFlightBatch.offsetState() != null || (isRecordLimitMode && (numRecordsAcquired > numRecordsRemaining))` and let `acquireSubsetBatchRecords` do the processing.

apoorvmittal10 reviewed (changes requested):
Thanks for the changes, some high level comments on the approach in SharePartition.

JimmyWang6 committed:
address comments

JimmyWang6 committed:
add more tests

JimmyWang6 committed:
add test when acknowledge in middle

JimmyWang6 committed:
fix ut

JimmyWang6 commented:
@adixitconfluent 
Thanks for review — updates are done:

1. a. The case should be covered by the following tests:
`testAcquireMultipleRecordsWithOverlapAndNewBatchInRecordLimitMode`
`testAcquireCachedStateInitialGapOverlapsWithActualPartitionGapInRecordLimitMode`
b.  `testAcquireWhenInsufficientRecordsInRecordLimitMode` — added.
c.  `testAcquireAndAcknowledgeMultipleSubsetRecordInRecordLimitMode`  — added.

3. Done.

JimmyWang6 committed:
nit

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java:
```diff
@@ -0,0 +1,92 @@
+
+    @Override
+    public String toString() {
+        return "ShareAcquireMode{"  +
```
nit: Because this is an enum, it's not really helpful to think of each instance as an object with two fields. I suggest something like `return "ShareAcquireMode(" + name + " (" + id + "));`.

JimmyWang6 replied:
Thanks for comment, fixed.

AndrewJSchofield reviewed (changes requested):
Thanks for the update. I've left another comment and will review some more later.

JimmyWang6 committed:
merge trunk

JimmyWang6 committed:
address comment

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -846,25 +851,35 @@ public ShareAcquiredRecords acquire(
-                AcquisitionLockTimerTask acquisitionLockTimeoutTask = scheduleAcquisitionLockTimeout(memberId, inFlightBatch.firstOffset(), inFlightBatch.lastOffset());
-                // Set the acquisition lock timeout task for the batch.
-                inFlightBatch.updateAcquisitionLockTimeout(acquisitionLockTimeoutTask);
+                long recordsToAcquiredCount = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
```
nit: `recordsToAcquireCount` sounds better instead of `recordsToAcquiredCount`

JimmyWang6 replied:
I think maybe we could use `numRecordsToAcquire`, what do you think?

adixitconfluent replied:
sounds good

JimmyWang6 replied:
Completed this one and also add a unit test to cover `filterShareAcquiredRecordsInRecordLimitMode`, will try to consider more scenarios later.

adixitconfluent commented:
Hi @JimmyWang6 , I don't think there is any unit test that covers `filterShareAcquiredRecordsInRecordLimitMode` (I ran the coverage on methods). Please write a few unit tests covering those scenarios.

AndrewJSchofield commented on server/src/test/java/org/apache/kafka/server/share/fetch/ShareFetchTest.java:
```diff
@@ -50,6 +50,7 @@ public class ShareFetchTest {
 
     private static final String GROUP_ID = "groupId";
     private static final String MEMBER_ID = "memberId";
+    private static final byte SHARE_ACQUIRE_MODE = 0;
```
This is the id for `BATCH_OPTIMIZED`. I would prefer to see `BATCH_OPTIMIZED` used as the parameter rather than `SHARE_ACQUIRE_MODE`.

JimmyWang6 replied:
Done.

AndrewJSchofield commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -5929,6 +5929,7 @@ class ReplicaManagerTest {
         Uuid.randomUuid.toString,
         future,
         topicPartitions,
+        0,
```
This is `ShareAcquireMode.BATCH_OPTIMIZED.id()` or similar I think. We should have a consistent way of referring to this constant in the broker code and tests.

JimmyWang6 replied:
Updated a couple of tests to make them consistent, thanks.

JimmyWang6 committed:
address comments

JimmyWang6 committed:
add test to cover filterShareAcquiredRecordsInRecordLimitMode

adixitconfluent commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -136,6 +137,7 @@ class ReplicaManagerTest {
   // Constants defined for readability
   private val partitionEpoch = 0
   private val brokerEpoch = 0L
+  private val batchOptimized = ShareAcquireMode.BATCH_OPTIMIZED.id();
```
nit: Given we only have a single test related to share groups in this file, I don't think we need a new variable for it, we can just use the value `ShareAcquireMode.BATCH_OPTIMIZED.id()` in the test where it is needed.

JimmyWang6 replied:
Done.

adixitconfluent commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -846,25 +851,35 @@ public ShareAcquiredRecords acquire(
-                // Set the acquisition lock timeout task for the batch.
-                inFlightBatch.updateAcquisitionLockTimeout(acquisitionLockTimeoutTask);
+                long recordsToAcquiredCount = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
+                int numRecordsRemaining = maxRecordsToAcquire - acquiredCount;
```
I don't think you addressed the comment https://github.com/apache/kafka/pull/20246/commits/8624ac5813d5881e523669b64c1b915ac09ca107#r2469546213, the variable name should be `numRecordsToAcquire` as we discussed in the comment thread

JimmyWang6 replied:
Oops, I think I accidentally missed something in my previous commit, but I've already fixed it.

adixitconfluent reviewed (changes requested):
@JimmyWang6 thanks for the commit.  Just a couple of comments.

JimmyWang6 committed:
fix nit

JimmyWang6 committed:
merge trunk

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1607,16 +1624,54 @@ private ShareAcquiredRecords acquireNewBatchRecords(
         }
     }
 
+    private AcquiredRecords filterShareAcquiredRecordsInRecordLimitMode(int maxFetchRecords, InFlightBatch inFlightBatch) {
```
I think we can get rid of this method if we change out approach.

JimmyWang6 replied:
@apoorvmittal10 I think you're right. I've just made the changes as you suggested, please take another look.

apoorvmittal10 reviewed (changes requested):
Thanks for the changes, high level approach comment for SharePartition. Once we align the changes in SharePartition, I ll review other files.

JimmyWang6 committed:
address review comments

JimmyWang6 committed:
merge trunk

JimmyWang6 committed:
fix build

JimmyWang6 commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -1518,7 +1526,6 @@ public void testMaybeInitializeAndAcquireWithMultipleBatches() {
         assertEquals(6, sharePartition.cachedState().get(5L).lastOffset());
         assertEquals(RecordState.ACQUIRED, sharePartition.cachedState().get(5L).batchState());
         assertEquals(1, sharePartition.cachedState().get(5L).batchDeliveryCount());
-        assertNull(sharePartition.cachedState().get(5L).offsetState());
```
Remove this assertion to resolve the Checkstyle error:
`SharePartitionTest.java:1464:5: Method testMaybeInitializeAndAcquireWithMultipleBatches length is 172 lines (max allowed is 170).`
When `offsetState` is non-null, `batchState` will be null, so calling
`sharePartition.cachedState().get(5L).batchDeliveryCount()` would throw an `IllegalStateException`.
Therefore, it is fine to remove this assertion.

JimmyWang6 committed:
remove unuse method

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java:
```diff
@@ -0,0 +1,89 @@
+            try {
+                of(acquireMode);
+            } catch (Exception e) {
+                throw new ConfigException(name, value, "Invalid value :" + acquireMode + " for configuration " +
```
Let's have
```
"Invalid value `" + acquireMode + "` for configuration "+
```
to match other similar classes such as AutoOffsetResetStrategy.java.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -816,8 +821,9 @@ public ShareAcquiredRecords acquire(
                 boolean fullMatch = checkForFullMatch(inFlightBatch, firstBatch.baseOffset(), lastOffsetToAcquire);
-
-                if (!fullMatch || inFlightBatch.offsetState() != null) {
+                long numRecordsToAcquire = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
```
The name should be `numRecordsInBatch` as we don't know yet if all records from the batch can be acquired or not.
```suggestion
                long numRecordsInBatch = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
```

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -816,8 +821,9 @@ public ShareAcquiredRecords acquire(
-                if (!fullMatch || inFlightBatch.offsetState() != null) {
+                long numRecordsToAcquire = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
+                int numRecordsRemaining = maxRecordsToAcquire - acquiredCount;
+                if (!fullMatch || inFlightBatch.offsetState() != null || (isRecordLimitMode && (numRecordsToAcquire > numRecordsRemaining))) {
```
May be:


```
         boolean recordLimitSubsetMatch = isRecordLimitMode ? checkForRecordLimitSubsetMatch(inflightBatch,.....) : false;
          if (!fullMatch || inFlightBatch.offsetState() != null || recordLimitSubsetMatch) {
          ....
 ```
```
         private boolean checkForRecordLimitSubsetMatch(InflightBatch inflightBatch, ....) {
                long numRecordsInBatch = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
                int numRecordsRemaining = maxRecordsToAcquire - acquiredCount;
                return numRecordsInBatch > numRecordsRemaining;
         }
```

JimmyWang6 replied:
Good idea. Modified as follows::
 
```java
boolean recordLimitSubsetMatch = isRecordLimitMode && checkForRecordLimitSubsetMatch(inFlightBatch, maxRecordsToAcquire, acquiredCount);
```

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1751,10 +1768,56 @@ private List<AcquiredRecords> createBatches(
                 .setLastOffset(lastAcquiredOffset)
                 .setDeliveryCount((short) 1));
 
-            result.forEach(acquiredRecords -> {
-                // Schedule acquisition lock timeout for the batch.
-                AcquisitionLockTimerTask timerTask = scheduleAcquisitionLockTimeout(memberId, acquiredRecords.firstOffset(), acquiredRecords.lastOffset());
-                // Add the new batch to the in-flight records along with the acquisition lock timeout task for the batch.
+            if (isRecordLimitMode) {
+                AcquiredRecords acquiredRecords = result.get(0);
+                if (acquiredRecords.lastOffset() - acquiredRecords.firstOffset() + 1 > maxFetchRecords) {
+                    InFlightBatch inFlightBatch = new InFlightBatch(
```
Can you please write comments in the code, explaing why we need the additional check and why only first element from list is compared.

JimmyWang6 replied:
Done.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1751,10 +1768,56 @@ private List<AcquiredRecords> createBatches(
+                }
+            } else {
+                addBatches(memberId, result);
+            }
```
If has a return statement anyways
```suggestion
            addBatches(memberId, result);
```

JimmyWang6 replied:
Done, thanks for comment.

apoorvmittal10 reviewed (changes requested):
Thanks for the changes, the changes in SharePartition are looking good now. I have some suggestions, prior reviewing test files.

JimmyWang6 committed:
address comments

JimmyWang6 committed:
fix nit

JimmyWang6 committed:
fix nit

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/requests/ShareFetchRequest.java:
```diff
@@ -61,6 +61,7 @@ public static Builder forConsumer(String groupId, ShareRequestMetadata metadata,
             data.setMaxBytes(maxBytes);
             data.setMaxRecords(maxRecords);
             data.setBatchSize(batchSize);
+            data.setShareAcquireMode(shareAcquireMode);
```
Please add a version-dependent condition in the `build(short)` method similar to what we have there already to ensure that we don't try to send a share acquire mode to version < 2 brokers.

JimmyWang6 replied:
Finished, changed like this:
```java
if (version < 2) {
    // The v1 does not support AcknowledgeType RENEW.
    if (data.isRenewAck()) {
        throw new UnsupportedVersionException("The v1 ShareFetch does not support AcknowledgeType.RENEW");
    }
    // The v1 does not support ShareAcquireMode record_limit.
    if (data.shareAcquireMode() == ShareAcquireMode.RECORD_LIMIT.id()) {
        throw new UnsupportedVersionException("The v1 ShareFetch does not support ShareAcquireMode.RECORD_LIMIT");
    }
}
```

AndrewJSchofield commented on clients/src/main/resources/common/message/ShareFetchRequest.json:
```diff
@@ -44,6 +44,8 @@
       "about": "The optimal number of records for batches of acquired records and acknowledgements." },
     { "name": "IsRenewAck", "type": "bool", "versions": "2+", "default": "false",
       "about": "Whether Renew type acknowledgements present in AcknowledgementBatches." },
+    { "name": "ShareAcquireMode", "type": "int8", "versions": "2+", "default": "0", "ignorable": true,
```
I have a slight preference to put this right before `IsRenewAck` because it seems more logically groups with the size fields.

JimmyWang6 replied:
Done.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchConfig.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    public String toString() {
+        return "FetchConfig{" +
```
`ShareFetchConfig`

JimmyWang6 replied:
Good catch, done~

AndrewJSchofield reviewed (changes requested):
Apart from the few minor comments, I'm happy with this on the client.

I do want to reiterate that we should not be splitting batches on the broker unless we really have to, even in record_limit mode. The limit is a limit, not a target. We might have to split batches if a small number of records is requested and the available batch is too large. In the case where we can give back entire batches but are still below the limit, then the entire batches should be returned.

JimmyWang6 committed:
address comments

JimmyWang6 committed:
fix typo

adixitconfluent reviewed (approved):
Thanks @JimmyWang6 for the changes, looks good to me from broker changes POV. You should address pending comments from @AndrewJSchofield and @apoorvmittal10 and await for approval from them before merging.

JimmyWang6 commented:
Hi @AndrewJSchofield,
Thanks for your review and comments.

> I do want to reiterate that we should not be splitting batches on the broker unless we really have to

I completely agree! Splitting batches will indeed introduce additional overhead for both the server and the client.

However, from a high-level perspective, if the producer configures larger values for `linger.ms` and `batch.size`, a single batch may contain a large number of records (e.g., 10,000), which could optimize throughput. In this scenario, if a shared consumer sets a relatively low value for `maxFetchRecords` (e.g., 5) in `record_limit` mode, batch splitting will inevitably occur on the broker. To mitigate this overhead, we could recommend that users reduce both the `batch.size` and `linger.ms` values.

> In the case where we can give back entire batches but are still below the limit, then the entire batches should be returned

This is essentially what I am trying to achieve based on the current implementation. Batch splitting only occurs in the `acquireNewBatchRecords` method, primarily in the following cases:

1. No overlap with request offsets in the cache.
2. The inflight batch is not a full match and only  a subset of records should be acquired.
3. There are no enough records up to `maxFetchRecords` in `cacheState` so we need to acquire mode records from batches.

I will further investigate whether additional optimizations can be applied to the cases below.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -380,14 +380,25 @@ public class ConsumerConfig extends AbstractConfig {
+            " If set to <code>record_limit</code>, the number of records returned in each poll() will not exceed the value of <code>max.poll.records</code>." +
+            " If set to <code>batch_optimized</code>, the number of records returned in each poll() call may exceed <code>max.poll.records</code>" +
+            " to align with batch boundaries for optimization.";
+    public static final String DEFAULT_SHARE_ACQUIRE_MODE = "batch_optimized";
```
Can we use a constant for this, such as `ShareAcquireMode.BATCH_OPTIMIZED.name()`?

JimmyWang6 replied:
Done.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/requests/ShareFetchRequest.java:
```diff
@@ -143,6 +145,10 @@ public ShareFetchRequest build(short version) {
                     throw new UnsupportedVersionException("The v1 ShareFetch does not support AcknowledgeType.RENEW");
                 }
+                // The v1 does not support ShareAcquireMode record_limit.
+                if (data.shareAcquireMode() == ShareAcquireMode.RECORD_LIMIT.id()) {
```
Let's have `!= ShareAcquireMode.BATCH_OPTIMIZED.id()` to cover us for additional modes in the future and equivalent changes in the comment and exception message. V1 only supports batch_optimized in effect.

JimmyWang6 replied:
Done.

JimmyWang6 committed:
address comments

JimmyWang6 commented:
@AndrewJSchofield Thank you for your patience and preciseness, I made some improvements, plz take a look again.

apoorvmittal10 reviewed (commented):
Broker changes look good to me!

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/InFlightBatch.java:
```diff
@@ -188,6 +190,29 @@ public InFlightState startBatchStateTransition(RecordState newState, DeliveryCou
+     * an acquisition lock timeout task for each offset.
+     * (targetOffset to lastOffset]: initialize each offset to {@link RecordState#AVAILABLE}
+     * and do not schedule any timer task for these offsets.
+     */
```
Add @param for correct javadocs

JimmyWang6 replied:
Roger that.

JimmyWang6 committed:
merge trunk

JimmyWang6 committed:
address comment

apoorvmittal10 reviewed (approved):
Reviewed server changes, LGTM!

AndrewJSchofield commented:
@JimmyWang6 Checkstyle failure in latest build.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -380,14 +380,25 @@ public class ConsumerConfig extends AbstractConfig {
+            " If set to <code>record_limit</code>, the number of records returned in each poll() will not exceed the value of <code>max.poll.records</code>." +
+            " If set to <code>batch_optimized</code>, the number of records returned in each poll() call may exceed <code>max.poll.records</code>" +
+            " to align with batch boundaries for optimization.";
+    public static final String DEFAULT_SHARE_ACQUIRE_MODE = ShareAcquireMode.RECORD_LIMIT.name();
```
The default is `BATCH_OPTIMIZED` I think according to KIP-1206 :)

JimmyWang6 replied:
Sorry, something went wrong while addressing the comments.

AndrewJSchofield reviewed (changes requested):
Probably just one final comment.

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchConfig.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    public String toString() {
+        return "FetchConfig{" +
```
```suggestion
        return "ShareFetchConfig{" +
```

JimmyWang6 replied:
🤦Fixed

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchConfig.java:
```diff
@@ -91,7 +91,7 @@ public FetchConfig(ConsumerConfig config) {
     @Override
     public String toString() {
-        return "FetchConfig{" +
+        return "ShareFetchConfig{" +
```
```suggestion
        return "FetchConfig{" +
```

apoorvmittal10 reviewed (approved):
Final comments as well, seems we mistakenly switched.

JimmyWang6 committed:
fix build error and couples of comments

AndrewJSchofield commented:
@JimmyWang6 I think this PR is good to go, but unfortunately there are conflicts to resolve. I expect this is the final revision.

JimmyWang6 committed:
merge trunk

JimmyWang6 commented:
@AndrewJSchofield Conflicts resolved, PTAL.

chia7712 commented on clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public String toString() {
+        return "ShareAcquireMode(" + name + " (" + id + "))";
```
Should this return the `name` instead? That would be consistent with other enum classes used for configuration

chia7712 commented on clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java:
```diff
@@ -0,0 +1,90 @@
+
+    public final String name;
+
+    public final byte id;
```
Since a method called `id()` already exists, is the `public` modifier necessary here?

JimmyWang6 replied:
@chia7712 Thanks for your comment. I will submit a minor PR to address them later.

chia7712 commented on clients/src/main/java/org/apache/kafka/clients/consumer/ShareAcquireMode.java:
```diff
@@ -0,0 +1,90 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
```
Is the `org.apache.kafka.clients.consumer.internals` better package?

JimmyWang6 replied:
Hi @chia7712 , 
Sorry for the late response. I agree with your point here and the KIP needs to be updated as well.

apoorvmittal10 replied:
Well I was thinking that as this is a client configuration which has been exposed to customer hence keeping it in public package makes sense. @chia7712 Shouldn't we?

chia7712 replied:
>  this is a client configuration which has been exposed to customer hence keeping it in public package makes sense.

Making this class public brings additional maintenance overhead. Therefore, exposing just the constant strings should be adequate, similar to how we handle `ShareAcknowledgementMode` and `AutoOffsetResetStrategy`. If the class must be exposed, we should aim to avoid exposing any public members

chia7712 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1762,10 +1779,57 @@ private List<AcquiredRecords> createBatches(
+
+                    cachedState.put(acquiredRecords.firstOffset(), inFlightBatch);
+                    sharePartitionMetrics.recordInFlightBatchMessageCount(
+                        acquiredRecords.lastOffset() - acquiredRecords.firstOffset() + 1);
```
Should we use `lastOffset` rather than `acquiredRecords.lastOffset()` to update the metrics? otherwise, it may cause duplicate count?

JimmyWang6 replied:
Hmm I think we need to do a lot more work to make `inFlightBatchMessageCount` metrics more accurately, because this metric doesn't update when records transition between `AVAILABLE`、`ACKNOWLEDGED` and `ARCHIVED`. In other words, this metric is always aligned with batch boundaries by design. Therefore, I'm inclined to keep the current implementation. WDYT?

apoorvmittal10 replied:
@chirag-wadhwa5 can you review this comment please and advise if we need more work here?

chirag-wadhwa5 replied:
I agree with @JimmyWang6. As per the current implementation, a single batch could have different records in different states, acquired by different share consumers. When a batch is indeed broken because of a subset of records being acknowledged with a different type, the metric is not updated even then. In fact, a new value is recorded with the metric only when a new in-flight batch is being created. In this case, the new batch's boundary is between acquiredRecords.lastOffset() and acquiredRecords.firstOffset(); therefore, according to me, the current implementation is more sensible here.

chia7712 replied:
@JimmyWang6 @chirag-wadhwa5 thanks for your explanation. It appears that I overlooked the `acquiredRecords.setLastOffset(lastOffset);`, which makes my previous comment a bit strange :laughing:

chirag-wadhwa5 replied:
Yes, even I read the code wrong. The line `acquiredRecords.setLastOffset(lastOffset);`, should appear after the metric value is recorded

chia7712 replied:
> The line acquiredRecords.setLastOffset(lastOffset);, should appear after the metric value is recorded

Ah, I think you've identified the true issue :smile: @JimmyWang6 would you mind addressing it in #20973?

JimmyWang6 replied:
Roger that.

chia7712 reviewed (commented):
@JimmyWang6 thanks for this cool improvement. I leave some questions. Please take a look
